qiskit/transpiler/passes/mapping/noise_adaptive_layout.py

```diff
@@ -0,0 +1,246 @@
+pick the best available hardware cx to execute this edge.
+Iterative step:
+When an edge exists with one endpoint unmapped,
+map that endpoint to a location which allows
+maximum reliability for CNOTs with previously mapped qubits.
+In the end if there are unmapped qubits (which don't
+participate in any CNOT), map them to any available
+hardware qubit.
+
+Note: even though a 'layout' is not strictly a property of the DAG,
+in the transpiler architecture it is best passed around between passes by
+being set in `property_set`.
+"""
+
+import math
+import networkx as nx
+
+from qiskit.mapper import Layout
+from qiskit.transpiler.basepasses import AnalysisPass
+from qiskit.transpiler.exceptions import TranspilerError
+
+
+class NoiseAdaptiveLayout(AnalysisPass):
+    """
+    Chooses a noise-adaptive Layout based on current calibration
+    data for the backend.
+    """
+
+    def __init__(self, backend_prop):
+        """
+        Chooses a Noise Adaptive Layout
+
+        Args:
+            backend_prop (BackendProperties): backend properties object
+
+        Raises:
+            TranspilerError: if invalid options
+        """
+        super().__init__()
+        self.backend_prop = backend_prop
+        self.swap_graph = nx.DiGraph()
+        self.cx_errors = {}
+        self.readout_errors = {}
+        self.available_hw_qubits = []
+        self.gate_list = []
+        self.gate_cost = {}
+        self.swap_paths = {}
+        self.swap_costs = {}
+        self.prog_graph = nx.Graph()
+        self.qreg_to_id = {}
+        self.pending_program_edges = []
+        self.prog2hw = {}
+
+    def _initialize_backend_prop(self):
+        """
+        Extract readout and CNOT errors and compute swap costs.
+        """
+        backend_prop = self.backend_prop
+        for ginfo in backend_prop.gates:
+            if ginfo.gate == 'cx':
+                for item in ginfo.parameters:
+                    if item.name == 'gate_error':
+                        g_reliab = 1.0 - item.value
+                        break
+                    else:
+                        g_reliab = 1.0
+                swap_reliab = -math.log(pow(g_reliab, 3))
+                self.swap_graph.add_edge(ginfo.qubits[0], ginfo.qubits[1], weight=swap_reliab)
+                self.swap_graph.add_edge(ginfo.qubits[1], ginfo.qubits[0], weight=swap_reliab)
+                self.cx_errors[(ginfo.qubits[0], ginfo.qubits[1])] = g_reliab
+                self.gate_list.append((ginfo.qubits[0], ginfo.qubits[1]))
+        idx = 0
+        for q in backend_prop.qubits:
+            for nduv in q:
+                if nduv.name == 'readout_error':
+                    self.readout_errors[idx] = 1.0 - nduv.value
+                    self.available_hw_qubits.append(idx)
+            idx += 1
+        for edge in self.cx_errors:
+            self.gate_cost[edge] = self.cx_errors[edge] * self.readout_errors[edge[0]] *\
+                self.readout_errors[edge[1]]
```

Comment:
why is the `self.gate_cost` here penalized by the readout errors of the qubits? given that `self.gate_cost` is used in `_select_best_remaining_cx()` it seems it should only contain the cx cost (in which case I'm not sure it's different from `self.cx_error`).

Reply:
Ok, I looked a bit more. Seems like this is part of your heuristic. I think that's fine for now. My only hesitation is that the qubits may participate in the `cx` gate, but later be swapped elsewhere and not actually measured in that location. But I'm not sure how to account for that.

Reply:
Yes, it is part of the heuristic. The heuristic basically tries to find a subset of qubits that have good CNOT and readout errors. When it wants to allocate a new program CNOT (which has both qubits unmapped), it finds a good hardware CNOT gate which also has good readout error on both qubits.  When it is allocating a program CNOT which has one endpoint mapped, it considers the cost as the swap cost * readout of the newly allocated hardware qubit. 

In the optimization based methods, we can separately account for CX and readout errors and other factors like coherence time, 1Q errors etc. With a greedy heuristic, I think it is cleaner to club both CNOT and readout errors into a single cost function. 

As you rightly point out, these qubits may get swapped and the readout error may not be applicable. We can come up with a better pass that has some feedback between layout choices and swap choices.